### ISSUE TITLE
Hanging connections

### ISSUE THREAD

**Issue by ineiti** [dedis/cothority#331]
When testing the `cosid`, sometimes the connections hang. I suppose the following happens:

- host1 connects to host2
- host2 saves the connection
- host1 goes down and up again
- host2 tries to connect host1 through a stale connection

So if the connection is stale, there should be a closing and opening of the connection, again.

Ismail and/or Nicolas, please add a trace of your server (if it's running with debug-level 3 or more).

**Comment by ineiti** [comment]
Some work has been done on this, but it can still fail sometimes. Need to dig in further.

**Comment by ineiti** [comment]
I propose we put a timeout on the connections (like 1 hour) and then close the connections if nothing happened during that time. Does that sound reasonable?

**Comment by nikkolasg** [comment]
I think before doing that we should make sure the bug is still hapenning ? Because from the scenario you typed in the issue, it's not supposed to happen anymore since we made some careful changes to sda.Host. More precisely, we made sure that a host 1 which sends a message to a host 2, if the sending fails, then host 1 tries to connect again.

**Comment by ineiti** [comment]
It still does happen. Waiting for some time and then testing the connections gives:

Sorry for the long post. I `check`ed the group-toml-file. Obviously my cothority-server was down. I restarted it after the first three `check`s. There are five in total. As you can see, from check 1..2..3 there are more and more successes. So something is wrong ;)
Also check 4 doesn't pass directly (after restarting my server), only check 5 does.

So there is still _something_ wront (sic)...

1st check:
```
altcode$ cosi check -g Documents/go/src/gopkg.in/dedis/cosi.v0/dedis_group.toml 
[+] Checking the availability and responsiveness of the servers in the group...
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 78.46.227.60:2000_Ineiti's : Error 'Received an invalid repsonse.'
[+] Checking server(s) 95.143.172.241:62306_Daeinar's : Success
[+] Checking server(s) 192.33.210.8:6879_EPFL : Success
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 78.46.227.60:2000_Ineiti's : Error 'timeout on signing request'
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 5.135.161.91:2000_Nikkolasg's : Error 'Received an invalid repsonse.'
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 95.143.172.241:62306_Daeinar's : Success
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 192.33.210.8:6879_EPFL : Error 'timeout on signing request'
[+] Checking server(s) 192.33.210.8:6879_EPFL 5.135.161.91:2000_Nikkolasg's : Error 'timeout on signing request'
[+] Checking server(s) 185.26.156.40:61117_Ismail's 78.46.227.60:2000_Ineiti's : Error 'timeout on signing request'
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 185.26.156.40:61117_Ismail's : Error 'Received an invalid repsonse.'
[+] Checking server(s) 185.26.156.40:61117_Ismail's 95.143.172.241:62306_Daeinar's : Success
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's 192.33.210.8:6879_EPFL : Error 'timeout on signing request'
[+] Checking server(s) 192.33.210.8:6879_EPFL 185.26.156.40:61117_Ismail's : Error 'timeout on signing request'
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 95.143.172.241:62306_Daeinar's : Error 'Received an invalid repsonse.'
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 78.46.227.60:2000_Ineiti's : Error 'timeout on signing request'
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 192.33.210.8:6879_EPFL : Error 'Received an invalid repsonse.'
[+] Checking server(s) 192.33.210.8:6879_EPFL 78.46.227.60:2000_Ineiti's : Error 'timeout on signing request'
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 192.33.210.8:6879_EPFL : Error 'timeout on signing request'
[+] Checking server(s) 192.33.210.8:6879_EPFL 95.143.172.241:62306_Daeinar's : Error 'timeout on signing request'
```

2nd check:
```
[+] Checking the availability and responsiveness of the servers in the group...
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 78.46.227.60:2000_Ineiti's : Error 'Received an invalid repsonse.'
[+] Checking server(s) 95.143.172.241:62306_Daeinar's : Success
[+] Checking server(s) 192.33.210.8:6879_EPFL : Success
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 78.46.227.60:2000_Ineiti's : Error 'timeout on signing request'
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 5.135.161.91:2000_Nikkolasg's : Error 'Received an invalid repsonse.'
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 95.143.172.241:62306_Daeinar's : Success
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 192.33.210.8:6879_EPFL : Error 'timeout on signing request'
[+] Checking server(s) 192.33.210.8:6879_EPFL 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's 78.46.227.60:2000_Ineiti's : Error 'timeout on signing request'
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 185.26.156.40:61117_Ismail's : Error 'Received an invalid repsonse.'
[+] Checking server(s) 185.26.156.40:61117_Ismail's 95.143.172.241:62306_Daeinar's : Success
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's 192.33.210.8:6879_EPFL : Error 'timeout on signing request'
[+] Checking server(s) 192.33.210.8:6879_EPFL 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 95.143.172.241:62306_Daeinar's : Error 'Received an invalid repsonse.'
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 78.46.227.60:2000_Ineiti's : Error 'timeout on signing request'
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 192.33.210.8:6879_EPFL : Error 'Received an invalid repsonse.'
[+] Checking server(s) 192.33.210.8:6879_EPFL 78.46.227.60:2000_Ineiti's : Error 'timeout on signing request'
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 192.33.210.8:6879_EPFL : Error 'timeout on signing request'
[+] Checking server(s) 192.33.210.8:6879_EPFL 95.143.172.241:62306_Daeinar's : Success
```

3rd check:
```
[+] Checking the availability and responsiveness of the servers in the group...
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 78.46.227.60:2000_Ineiti's : Error 'Received an invalid repsonse.'
[+] Checking server(s) 95.143.172.241:62306_Daeinar's : Success
[+] Checking server(s) 192.33.210.8:6879_EPFL : Success
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 78.46.227.60:2000_Ineiti's : Error 'timeout on signing request'
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 5.135.161.91:2000_Nikkolasg's : Error 'Received an invalid repsonse.'
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 95.143.172.241:62306_Daeinar's : Success
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 192.33.210.8:6879_EPFL : Success
[+] Checking server(s) 192.33.210.8:6879_EPFL 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's 78.46.227.60:2000_Ineiti's : Error 'timeout on signing request'
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 185.26.156.40:61117_Ismail's : Error 'Received an invalid repsonse.'
[+] Checking server(s) 185.26.156.40:61117_Ismail's 95.143.172.241:62306_Daeinar's : Success
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's 192.33.210.8:6879_EPFL : Success
[+] Checking server(s) 192.33.210.8:6879_EPFL 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 95.143.172.241:62306_Daeinar's : Error 'Received an invalid repsonse.'
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 78.46.227.60:2000_Ineiti's : Error 'timeout on signing request'
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 192.33.210.8:6879_EPFL : Error 'Received an invalid repsonse.'
[+] Checking server(s) 192.33.210.8:6879_EPFL 78.46.227.60:2000_Ineiti's : Error 'timeout on signing request'
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 192.33.210.8:6879_EPFL : Success
[+] Checking server(s) 192.33.210.8:6879_EPFL 95.143.172.241:62306_Daeinar's : Success
```

4th check - restarted Ineiti's server
```
[+] Checking the availability and responsiveness of the servers in the group...
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 78.46.227.60:2000_Ineiti's : Success
[+] Checking server(s) 95.143.172.241:62306_Daeinar's : Success
[+] Checking server(s) 192.33.210.8:6879_EPFL : Success
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 78.46.227.60:2000_Ineiti's : Error 'timeout on signing request'
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 95.143.172.241:62306_Daeinar's : Success
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 192.33.210.8:6879_EPFL : Success
[+] Checking server(s) 192.33.210.8:6879_EPFL 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's 78.46.227.60:2000_Ineiti's : Success
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's 95.143.172.241:62306_Daeinar's : Success
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's 192.33.210.8:6879_EPFL : Success
[+] Checking server(s) 192.33.210.8:6879_EPFL 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 95.143.172.241:62306_Daeinar's : Success
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 78.46.227.60:2000_Ineiti's : Success
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 192.33.210.8:6879_EPFL : Success
[+] Checking server(s) 192.33.210.8:6879_EPFL 78.46.227.60:2000_Ineiti's : Success
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 192.33.210.8:6879_EPFL : Success
[+] Checking server(s) 192.33.210.8:6879_EPFL 95.143.172.241:62306_Daeinar's : Success
```

5th check:
```
[+] Checking the availability and responsiveness of the servers in the group...
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 78.46.227.60:2000_Ineiti's : Success
[+] Checking server(s) 95.143.172.241:62306_Daeinar's : Success
[+] Checking server(s) 192.33.210.8:6879_EPFL : Success
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 78.46.227.60:2000_Ineiti's : Success
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 95.143.172.241:62306_Daeinar's : Success
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 5.135.161.91:2000_Nikkolasg's 192.33.210.8:6879_EPFL : Success
[+] Checking server(s) 192.33.210.8:6879_EPFL 5.135.161.91:2000_Nikkolasg's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's 78.46.227.60:2000_Ineiti's : Success
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's 95.143.172.241:62306_Daeinar's : Success
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 185.26.156.40:61117_Ismail's 192.33.210.8:6879_EPFL : Success
[+] Checking server(s) 192.33.210.8:6879_EPFL 185.26.156.40:61117_Ismail's : Success
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 95.143.172.241:62306_Daeinar's : Success
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 78.46.227.60:2000_Ineiti's : Success
[+] Checking server(s) 78.46.227.60:2000_Ineiti's 192.33.210.8:6879_EPFL : Success
[+] Checking server(s) 192.33.210.8:6879_EPFL 78.46.227.60:2000_Ineiti's : Success
[+] Checking server(s) 95.143.172.241:62306_Daeinar's 192.33.210.8:6879_EPFL : Success
[+] Checking server(s) 192.33.210.8:6879_EPFL 95.143.172.241:62306_Daeinar's : Success
```

**Comment by nikkolasg** [comment]
So the first 3 checks are not showing anything here, right ? 
But the fact that it did not worked immediately just after you restarted your server is showing something yes. And having a counter here won't change the problem unfortunately.

**Comment by ineiti** [comment]
As the check 1-3 don't fail on the same hosts I think there is some kind of `recovery` of the system, but not enough... I'll implement some timeout in `network.go` so that connections that are don't used for more than 10 minutes get closed.

**Comment by ineiti** [comment]
Should be fixed with #523
